types/src/validator_verifier.rs

```diff
@@ -79,77 +84,96 @@ impl ValidatorConsensusInfo {
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
 pub struct ValidatorVerifier {
-    /// An ordered map of each validator's on-chain account address to its pubkeys
-    /// and voting power.
-    address_to_validator_info: BTreeMap<AccountAddress, ValidatorConsensusInfo>,
+    /// A vector of each validator's on-chain account address to its pubkeys and voting power.
+    validator_infos: Vec<ValidatorConsensusInfo>,
     /// The minimum voting power required to achieve a quorum
-    quorum_voting_power: u64,
+    #[serde(skip)]
+    quorum_voting_power: u128,
     /// Total voting power of all validators (cached from address_to_validator_info)
-    total_voting_power: u64,
+    #[serde(skip)]
+    total_voting_power: u128,
+    /// In-memory index of account address to its index in the vector, does not go through serde.
+    #[serde(skip)]
+    address_to_validator_index: HashMap<AccountAddress, usize>,
+}
+
+/// Reconstruct fields from the raw data upon deserialization.
+impl<'de> Deserialize<'de> for ValidatorVerifier {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(rename = "ValidatorVerifier")]
+        struct RawValidatorVerifier {
+            validator_infos: Vec<ValidatorConsensusInfo>,
+        }
+
+        let RawValidatorVerifier { validator_infos } =
+            RawValidatorVerifier::deserialize(deserializer)?;
+
+        Ok(ValidatorVerifier::new(validator_infos))
+    }
 }
 
 impl ValidatorVerifier {
+    /// Private constructor to calculate the in-memory index
+    fn _new(
+        validator_infos: Vec<ValidatorConsensusInfo>,
+        quorum_voting_power: u128,
+        total_voting_power: u128,
+    ) -> Self {
+        let address_to_validator_index = validator_infos
+            .iter()
+            .enumerate()
+            .map(|(index, info)| (info.address, index))
+            .collect();
+        Self {
+            validator_infos,
+            quorum_voting_power,
+            total_voting_power,
+            address_to_validator_index,
+        }
+    }
+
     /// Initialize with a map of account address to validator info and set quorum size to
     /// default (`2f + 1`) or zero if `address_to_validator_info` is empty.
-    pub fn new(
-        address_to_validator_info: BTreeMap<AccountAddress, ValidatorConsensusInfo>,
-    ) -> Self {
-        let total_voting_power = sum_voting_power(&address_to_validator_info);
-        let quorum_voting_power = if address_to_validator_info.is_empty() {
+    pub fn new(validator_infos: Vec<ValidatorConsensusInfo>) -> Self {
+        let total_voting_power = sum_voting_power(&validator_infos);
+        let quorum_voting_power = if validator_infos.is_empty() {
             0
         } else {
             total_voting_power * 2 / 3 + 1
```

Comment:
Unrelated to this change, but will this cause a potential overflow?

Reply:
it's sum from u64, so it can't really overflow